shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,173 @@
+import 'dart:async';
+import 'dart:ffi';
+import 'dart:isolate';
+
+import 'package:ffi/ffi.dart';
+import 'package:meta/meta.dart';
+import 'package:shorebird_code_push/src/generated/updater_bindings.g.dart';
+import 'package:shorebird_code_push/src/shorebird_updater.dart';
+import 'package:shorebird_code_push/src/updater.dart';
+
+@visibleForTesting
+
+/// Type definition for [Isolate.run].
+typedef IsolateRun = Future<R> Function<R>(
+  FutureOr<R> Function(), {
+  String? debugName,
+});
+
+/// {@template shorebird_updater_io}
+/// The Shorebird IO Updater.
+/// {@endtemplate}
+class ShorebirdUpdaterImpl implements ShorebirdUpdater {
+  /// {@macro shorebird_updater_io}
+  ShorebirdUpdaterImpl(this._updater, {IsolateRun? run})
+      : _run = run ?? Isolate.run {
+    try {
+      // If the Shorebird Engine is not available, this will throw an exception.
+      // FIXME: Run this in an isolate or refactor the updater to avoid risking
+      // a hang.
+      _updater.currentPatchNumber();
+      _isAvailable = true;
+    } catch (_) {
+      logShorebirdEngineUnavailableMessage();
+      _isAvailable = false;
+    }
+  }
+
+  late final bool _isAvailable;
+
+  final Updater _updater;
+
+  final IsolateRun _run;
+
+  @override
+  bool get isAvailable => _isAvailable;
+
+  @override
+  Future<Patch?> readCurrentPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.currentPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
```

Comment:
I might have abstracted this into a _readPatch(_updater.currentPatchNumber) call to avoid the duplication (even if the API is still "duplicated" the internals arent).